oktapam/resource_database.go

```diff
@@ -0,0 +1,380 @@
+			},
+		},
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceDatabaseReadImport,
+		},
+	}
+}
+
+func resourceDatabaseCreate(ctx context.Context, d *schema.ResourceData, m any) diag.Diagnostics {
+	c := getSDKClientFromMetadata(m)
+	resourceGroupID := d.Get(attributes.ResourceGroup).(string)
+	projectID := d.Get(attributes.Project).(string)
+	canonicalName := d.Get(attributes.CanonicalName).(string)
+	dbType := d.Get(attributes.DatabaseType).(string)
+
+	selectorLabels, err := GetMapPtrFromResource[string](attributes.ManagementGatewaySelector, d)
+	if err != nil {
+		return err
+	}
+
+	mgmtType, mgmtDetails, err := mgmtConnectionDetailsFromResource(ctx, c, d)
+	if err != nil {
+		return err
+	}
+
+	if createdDb, err := client.CreateDatabase(ctx, c, resourceGroupID, projectID, canonicalName, dbType, mgmtType, *mgmtDetails, selectorLabels); err != nil {
+		return diag.FromErr(err)
+	} else if createdDb == nil {
+		d.SetId("")
+	} else {
+		d.SetId(createdDb.Id)
+	}
+
+	return resourceDatabaseRead(ctx, d, m)
+}
+
+func resourceDatabaseRead(ctx context.Context, d *schema.ResourceData, m any) diag.Diagnostics {
+	c := getSDKClientFromMetadata(m)
+	resourceGroupID := d.Get(attributes.ResourceGroup).(string)
+	projectID := d.Get(attributes.Project).(string)
+	databaseID := d.Get(attributes.ID).(string)
+
+	database, err := client.GetDatabase(ctx, c, resourceGroupID, projectID, databaseID)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	wrap := wrappers.DatabaseResourceResponseWrapper{*database}
+	attributeOverrides := utils.GenerateAttributeOverrides(d, wrap)
+	for key, value := range wrap.ToResourceMap(attributeOverrides) {
+		if err := d.Set(key, value); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return nil
+}
+
+func resourceDatabaseUpdate(ctx context.Context, d *schema.ResourceData, m any) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	c := getSDKClientFromMetadata(m)
+	dbID := d.Get(attributes.ID).(string)
+	resourceGroupID := d.Get(attributes.ResourceGroup).(string)
+	projectID := d.Get(attributes.Project).(string)
+	canonicalName := d.Get(attributes.CanonicalName).(string)
+	dbType := d.Get(attributes.DatabaseType).(string)
+
+	selectorLabels, err := GetMapPtrFromResource[string](attributes.ManagementGatewaySelector, d)
+	if err != nil {
+		return err
+	}
+
+	mgmtType, mgmtDetails, err := mgmtConnectionDetailsFromResource(ctx, c, d)
+	if err != nil {
+		return err
+	}
+
+	if dbID == "" {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "could not obtain database id from resource",
+		})
+	}
+	if resourceGroupID == "" {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "could not obtain resource group id from resource",
+		})
+	}
+	if projectID == "" {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "could not obtain project id from resource",
+		})
+	}
+
+	if diags != nil {
```

Comment:
You don't need to check all the above error conditions related to dbID, resourceGroupID, projectID. These attributes are mandatory in the resource schema, so while updating the same resource these must be present. I should have put this comment in the earlier feedback.

Reply:
and im not sure why i didnt remove them last night.